tracer/src/Datadog.Tracer.Native/fault_tolerant_rewriter.cpp

```diff
@@ -0,0 +1,594 @@
+    }
+    else if (sizeof(UINT_PTR) == 8) // 64-bit
+    {
+        rewriterWrapper.LoadInt64(static_cast<INT64>(moduleId));
+    }
+
+    rewriterWrapper.LoadInt32(methodId);
+    rewriterWrapper.LoadStr(instrumentationVersionIdToken);
+    rewriterWrapper.LoadInt32(static_cast<INT32>(instrumentingProduct));
+    ILInstr* reportSuccessfulInstrumentation;
+    faultTolerantTokens->WriteReportSuccessfulInstrumentation(&rewriterWrapper, &reportSuccessfulInstrumentation);
+
+    hr = rewriter.Export();
+
+    if (IsDumpILRewriteEnabled())
+    {
+        std::string original_code = m_corProfiler->GetILCodes(
+            "*** FaultTolerantRewriter::InjectSuccessfulInstrumentation() Original Code: ", &rewriter, *caller,
+            moduleHandler->GetModuleMetadata()->metadata_import);
+        Logger::Info(original_code);
+    }
+
+    if (FAILED(hr))
+    {
+        Logger::Warn("Failed to emit IL for injecting successful instrumentation, ModuleID=", moduleId);
+        return hr;
+    }
+
+    Logger::Info("Successfully added injection of successful instrumentation report, moduleId = ", moduleId);
+    return hr;
+}
+
+HRESULT FaultTolerantRewriter::RewriteInternal(RejitHandlerModule* moduleHandler,
+                                               RejitHandlerModuleMethod* methodHandler,
+                                               ICorProfilerFunctionControl* pFunctionControl)
+{
+    if (!is_fault_tolerant_instrumentation_enabled)
+    {
+        return m_methodRewriter->Rewrite(moduleHandler, methodHandler, pFunctionControl);
+    }
+
+    const auto moduleId = moduleHandler->GetModuleId();
+    const auto methodId = methodHandler->GetMethodDef();
+
+    if (FaultTolerantTracker::Instance()->IsKickoffMethod(moduleId, methodId))
+    {
+        const auto instrumentationVersion = m_methodRewriter->GetInstrumentationVersion(moduleHandler, methodHandler);
+        const auto instrumentingProduct = m_methodRewriter->GetInstrumentingProduct(moduleHandler, methodHandler);
+
+        if (FaultTolerantTracker::Instance()->IsInstrumentationVersionSucceeded(moduleId, methodId, instrumentationVersion, instrumentingProduct))
+        {
+            // Request Revert for the original method
+            //std::vector<MethodIdentifier> requests = {{moduleId, methodId}};
+            //auto promise = std::make_shared<std::promise<void>>();
+            //auto future = promise->get_future();
+            //m_rejit_handler->EnqueueRequestRevert(requests, promise);
+            //future.get();
+
+            // We also set the original IL function body, in case the revert did not succeeded.
+            // We do that just as a safeguard, it has harmless effect.
+            //const auto [pMethodBytes, methodSize] = FaultTolerantTracker::Instance()->GetILBodyAndSize(moduleId, methodId);
+            //pFunctionControl->SetILFunctionBody(methodSize, pMethodBytes);
+
+            return m_methodRewriter->Rewrite(moduleHandler, methodHandler, pFunctionControl);
```

Comment:
Can you confirm that my understanding is correct on the sequence of events to get here? Here's the timeline I have:
1. We rewrite the method once and the kickoff instrumentation is generated through the else branch
2. The "Instrumented" was entered successfully so `ReportSuccessfulInstrumentation` was called, which invokes a `Revert` and then a `Rejit`
3. We get to this branch as a result of the `Rejit`

Is it also correct that this rewrite will end up rewriting the method so there's no more Kickoff, so we directly invoke the instrumentations in this method body?

Reply:
Exactly! I'm basically holding an `instrumentation version` (a string) that is embedded into the kickoff and the instrumentation version. If we succeeded in the instrumentation, `ReportSuccessfulInstrumentation` is getting with the `instrumentation version`. Upon landing in this code block again, meaning after rejitting, we'll skip the kickoff entirely and perform the instrumentation directly.
The underlying rewriter, which is in use to get this `instrumentation version`, has the responsibility to return the same string for the same bytecode instrumentation to be performed.
- In the Debugger case, I'm building this 'instrumentation version` based on the probes that exist on that method.
- In the Tracer case, I presume you can return the same string in all cases since the bytecode instrumentation is not affected by external configuration.

Does that make sense?

Reply:
Makes sense to me, thanks!